File: godfather/roles/mafia/goon.py
from godfather.roles.mixins import SingleAction, Shooter, MafiaMember

DESCRIPTION = 'You may shoot someone every night.'


class Goon(MafiaMember, Shooter, SingleAction):
    """
    A member of organized crime, trying to work their way to the top.

    - Win Condition: Kill anyone that will not submit to the Mafia.

    + Abilities:
    + You can choose to attack if the Godfather doesn't give orders.
    + If the Godfather dies, you will become the next Godfather.
    """
    name = 'Goon'
    description = DESCRIPTION
    unique = True

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.categories.append('Mafia Killing')

    async def on_pm_command(self, ctx, game, player, args):
        if any(gf_action := list(filter(lambda action: action['player'].role.name == 'Godfather', game.night_actions))):
            gf_action = gf_action[0]
            text = 'The Godfather has ordered you to stay home' \
                if gf_action['action'] is None \
                else 'The Godfather has ordered you to shoot {}'.format(gf_action['target'].user.name)
            return await ctx.send(text)
        await super().on_pm_command(ctx, game, player, args)

    async def tear_down(self, actions, player, target):

        record = actions.record[target.user.id]['nightkill']
        success = record['result'] and player in record['by']

        if not success:
            return await player.user.send('Your target was too strong to kill!')
        await target.user.send('You were shot by a Goon. You have died!')

    def can_do_action(self, game):
        if game.setup.name == 'dethy' and game.cycle == 1:
            return False, 'You cannot shoot N1 in Dethy.'
        return super().can_do_action(game)

File: src/lib/mafia/roles/mafia/Goon.ts
import MafiaRole from '@mafia/mixins/MafiaRole';
import type Player from '@mafia/structures/Player';
import type { Message } from 'discord.js';
import KillerAction from '@mafia/actions/common/KillerAction';
import type { NightAction } from '@mafia/managers/NightAction';
import { Attack } from '@mafia/managers/NightActionsManager';
import { ActionRole } from '@mafia/structures/ActionRole';

class Goon extends ActionRole {
	public name = 'Goon';
	public actions: NightAction[] = [new KillerAction(this, Attack.Basic, 'shoot')];

	public constructor(player: Player) {
		super(player);
		this.description = this.game.t('roles/mafia:goonDescription');
	}

	public async onPmCommand(message: Message, command: string, ...args: string[]) {
		const gfAction = this.game.nightActions.find((action) => action.actor.role.name === 'Godfather');
		if (gfAction) {
			return message.channel.send(
				gfAction.action === undefined
					? this.game.t('roles/mafia:goonMessageHome')
					: this.game.t('roles/mafia:goonMessageShoot', { target: gfAction.target! as Player })
			);
		}
		return super.onPmCommand(message, command, ...args);
	}

	public canUseAction() {
		if (this.game.setup!.name === 'dethy' && this.game.cycle === 1) {
			return { check: false, reason: this.game.t('roles/mafia:goonDethy') };
		}
		return super.canUseAction();
	}

	public static unique = true;
}

Goon.categories = [...Goon.categories, 'Mafia Killing'];
Goon.aliases = ['Mafioso'];

export default MafiaRole(Goon);
